extensions/amp-instagram/1.0/instagram.js

```diff
@@ -0,0 +1,82 @@
+/**
+ * Copyright 2020 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import * as Preact from '../../../src/preact';
+import * as styles from './instagram.css';
+import {ContainWrapper} from '../../../src/preact/component';
+import {getData} from '../../../src/event-helper';
+import {px, resetStyles, setStyle, setStyles} from '../../../src/style';
+import {useMountEffect} from '../../../src/preact/utils';
+
+/**
+ * @param {!InstagramProps} props
+ * @return {PreactDef.Renderable}
+ */
+export function Instagram({shortcode, captioned, width, layout, style, alt}) {
+  captioned = captioned ? 'captioned/' : '';
+  const iframeRef = Preact.useRef(null);
+  const [heightStyle, setHeightStyle] = Preact.useState(null);
+  const [opacityStyle, setOpacityStyle] = Preact.useState(0);
+  const widthStyle = layout == 'responsive' ? '100%' : px(String(width));
+
+  useMountEffect(() => {
+    function handleMessage(event) {
+      if (
+        event.origin != 'https://www.instagram.com' ||
+        event.source != iframeRef.current.contentWindow
+      ) {
+        return;
+      }
+
+      const data = JSON.parse(getData(event));
+
+      if (data['type'] == 'MEASURE') {
+        setHeightStyle(String(data['details']['height']) + 'px');
```

Comment:
1. Let's set this height style as an object so we can spread `...` as below.
2. No need to cast to string. If `data['details']['height']` comes as a number, Preact knows what to do with it when it's `height`.

```suggestion
        setHeightStyle({'height': data['details']['height']});
```

3. We may not need this (@dvoytenko) since `amp-instagram` already does [`forceChangeHeight`](https://github.com/ampproject/amphtml/blob/b6313e372fdd1298928e2417dcc616b03288e051/extensions/amp-instagram/0.1/amp-instagram.js#L226), but we might like to take an optional `resize` callback prop to override from AMP's side. If so:

```suggestion
        if (typeof resize === 'function') {
          resize();
        } else {
          setHeightStyle({'height': data['details']['height']});
        }
```

Reply:
I'm going to save this change to discuss for the end of the PR after everything else is completed

Reply:
Confirmed that you need to `forceChangeHeight` due to nested structure:

```js
class AmpInstagram extends PreactBaseElement {
  /** @override */
  init() {
    return dict({
      'resize': (height) => {
        this.forceChangeHeight(height);
      },
    });
  }
  // ...
}
```